modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+
+[NOTE]
+ The use of constraints is optional, but constraints can be useful in big scripts or multiple scripts. The type system can help you avoid bugs in your DataWeave logic because it causes the type-checking algorithm to run.
+
+The types supported by the DataWeave type system are divided into three categories:
+
+* Simple Types
+* Composite Types
+* Complex Types
+
+== Simple Types
+
+Simple types represent values such as strings, Booleans, and so on. These values are atomic; they are not composed of other values. These are the simple types:
+
+* `String`
+* `Boolean`
+* `Number`
+* `Regex`
+* `Null`
+* Temporal: `Date`, `DateTime`, `LocalDateTime`, `LocalTime`, `Time`, `Period`
+//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType
+
+=== Null
+
+`Null` is a type of only one value, the value `null`, which means that `null` cannot be assigned to the type `String` or any other type except for the type `Null`.
+
+For example, the `repeat` function in the String module has the signature `repeat(String, Number): String`, which means that the function only accepts as a first parameter a value of type `String`, and as a second parameter a value of type `Number.` The function returns a value of type `String`.
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat("a", 3)
+----
+
+This returns `&quot;aaa&quot;`, but:
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(null, 3)
+----
+
+Throws an error because the value `null` cannot be assigned to the parameter that expects a value of type String.
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(Null, 3)
+----
+
+The script assigns to the first parameter a value that is the type `Null` (the type of this value
+is `Type&lt;Null&gt;`), and it returns `&quot;NullNullNull&quot;` because there is an autocoercion that coerces the value to the type `String` (`&quot;Null&quot;`).
+
+Some other functions have an overloaded signature for the function (the function has multiples definitions for different parameter types) that let you call it with `null`.
+
+For example, the `isNumeric` function from `dw::core::Strings` has the signature `isNumeric(String): Boolean` (receives a string and successfully returns a result if that string is numeric), but the function is also defined as `isNumeric(Null): Boolean`. +
+
+So, if you call the function `isNumeric` with `null` (`isNumeric(null)`), the function does not throw an exception. At runtime, the function that is dispatching algorithm selects the correct function based on the types of the values you use. The function returns `false` because of the definition of `isNumeric` for `null` as the argument.
+
+== Composite Types
+
+A composite types contains other values. The composite types are `Array`, `Object` and `Function`.
```

Comment:
It contains values other than what?
Do we mean that it contains more than one value?
Or do we mean that it contains values other than `null`?
Also...:
```suggestion
A composite type contains other values. The composite types are `Array`, `Object`, and `Function`.
```